// wavesurfer.js/dist/base-plugin.js
import EventEmitter from './event-emitter.js';
/** Base class for wavesurfer plugins */
/**
 * BasePlugin - Abstract base class for all WaveSurfer plugins
 * 
 * This class provides a common foundation for creating WaveSurfer plugins by:
 * - Extending EventEmitter to enable event-driven communication
 * - Managing plugin lifecycle (initialization, destruction)
 * - Handling subscriptions to prevent memory leaks
 * - Providing a standardized interface for plugin development
*/
export class BasePlugin extends EventEmitter {
    /** Create a plugin instance */
    constructor(options) {
        // Call the parent EventEmitter constructor to initialize event handling capabilities
        super();
        // Initialize array to store event subscription cleanup functions
        // Used to prevent memory leaks when the plugin is destroyed
        this.subscriptions = [];
        // Store the plugin configuration options for use throughout the plugin lifecycle
        this.options = options;
    }
    /** Called after this.wavesurfer is available */
    onInit() {
        return;
    }
    /** Do not call directly, only called by WavesSurfer internally */
    _init(wavesurfer) {
        this.wavesurfer = wavesurfer;
        this.onInit();
    }
    /** Destroy the plugin and unsubscribe from all events */
    //Performs cleanup to prevent memory leaks and properly teardown the plugin
    destroy() {
        // Emit 'destroy' event to notify listeners that the plugin is being destroyed
        // This allows other components to perform their own cleanup if needed
        this.emit('destroy');
        // Execute all stored unsubscribe functions to clean up event listeners
        // This prevents memory leaks by removing all event subscriptions
        this.subscriptions.forEach((unsubscribe) => unsubscribe());
    }
}
export default BasePlugin;
